lib/elixir/src/elixir_bitstring.erl

```diff
@@ -127,6 +128,31 @@ validate_bit_type_args(Meta, unit, [Unit], E) when not is_integer(Unit) ->
 validate_bit_type_args(_Meta, _Expr, _Args, _E) ->
   ok.
 
+validate_bit_type_if_literal_bin(Left, Meta, Type, E) when is_binary(Left) ->
+  case valid_bit_type_for_literal_bin(Type) of
+    true ->
+      ok;
+    false ->
+      Message =
+        "invalid types for literal string in <<>>. Accepted types are: "
+        "little, big, utf8, utf16, utf32, bits, bytes, binary, bitstring",
+      elixir_errors:compile_error(Meta, ?m(E, file), Message)
+  end;
+validate_bit_type_if_literal_bin(_Left, _Meta, _Type, _E) ->
+  ok.
+
+valid_bit_type_for_literal_bin(little) -> true;
+valid_bit_type_for_literal_bin(big) -> true;
+valid_bit_type_for_literal_bin(utf8) -> true;
+valid_bit_type_for_literal_bin(utf16) -> true;
+valid_bit_type_for_literal_bin(utf32) -> true;
+valid_bit_type_for_literal_bin(bytes) -> true;
+valid_bit_type_for_literal_bin(binary) -> true;
+valid_bit_type_for_literal_bin(bits) -> true;
+valid_bit_type_for_literal_bin(bitstring) -> true;
+valid_bit_type_for_literal_bin(default) -> true;
+valid_bit_type_for_literal_bin(_) -> false.
```

Comment:
We could strip `validate_bit_type_for_literal_bin/1` down a bit if we reused `types_allow_splice/1`, but I think this reads clearer and matches exactly all the types we mention in the error message above (+ `default`), and conceptually is easy to follow ("these are the types you can use with literal binaries in `<<>>`").